mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,431 @@
+struct BlockHeader {
+    block_idx: BlockIndex,
+    etag: String,
+    s3_key: String,
+    header_checksum: u32,
+}
+
+#[derive(Debug)]
+enum BlockHeaderError {
+    /// The header's fields or its checksum is corrupt
+    IntegrityError,
+    /// One or more of the fields in this header were incorrect
+    FieldMismatchError,
+}
+
+impl From<BlockHeaderError> for DataCacheError {
+    fn from(value: BlockHeaderError) -> Self {
+        match value {
+            BlockHeaderError::IntegrityError | BlockHeaderError::FieldMismatchError => {
+                DataCacheError::InvalidBlockContent
+            }
+        }
+    }
+}
+
+impl BlockHeader {
+    /// Creates a new [BlockHeader]
+    pub fn new(block_idx: BlockIndex, etag: String, s3_key: String) -> Self {
+        let header_checksum = Self::compute_checksum(block_idx, &etag, &s3_key).value();
+        BlockHeader {
+            block_idx,
+            etag,
+            s3_key,
+            header_checksum,
+        }
+    }
+
+    fn compute_checksum(block_idx: BlockIndex, etag: &str, s3_key: &str) -> Crc32c {
+        let mut hasher = crc32c::Hasher::new();
+        hasher.update(block_idx.to_be_bytes().as_ref());
+        hasher.update(etag.as_bytes());
+        hasher.update(s3_key.as_bytes());
+        hasher.finalize()
+    }
+
+    /// Validate the integrity of the contained data.
+    ///
+    /// Execute this method before acting on the data contained within.
+    pub fn validate(&self, s3_key: &str, etag: &str, block_idx: BlockIndex) -> Result<(), BlockHeaderError> {
+        let s3_key_match = s3_key == self.s3_key;
+        let etag_match = etag == self.etag;
+        let block_idx_match = block_idx == self.block_idx;
+
+        if s3_key_match && etag_match && block_idx_match {
+            if Self::compute_checksum(block_idx, etag, s3_key).value() != self.header_checksum {
+                Err(BlockHeaderError::IntegrityError)
+            } else {
+                Ok(())
+            }
+        } else {
+            error!(
+                s3_key_match,
+                etag_match, block_idx_match, "block data did not match expected values",
+            );
+            Err(BlockHeaderError::FieldMismatchError)
+        }
+    }
+}
+
+/// Represents a fixed-size chunk of data that can be serialized.
+#[derive(Serialize, Deserialize, Debug)]
+struct DataBlock {
+    /// Information describing the content of `data`, to be used to verify correctness
+    header: BlockHeader,
+    /// Cached bytes
+    data: Bytes,
+    /// Checksum over the bytes in `data`
+    data_checksum: u32,
+}
+
+impl DataBlock {
+    /// Create a new [DataBlock].
+    ///
+    /// This may return an integrity error if the checksummed byte buffer is found to be corrupt.
+    /// However, this check is not guaranteed and it shouldn't be assumed that the data within the block is not corrupt.
+    fn new(cache_key: CacheKey, block_idx: BlockIndex, bytes: ChecksummedBytes) -> DataCacheResult<Self> {
```

Comment:
nit: both here and in `data()`, it seems early to map errors in `DataCacheError`s, even more without propagating the root cause. But can be reviewed/addressed later.

Reply:
I'll update at least to help the error types make more sense. I think they need to be split in two. I'll defer mapping the errors too.